### PR TITLE
py-torch-nvidia-apex: Add 24.04.01 and variants from the readme

### PR DESCRIPTION
**Pull request by aurianer**
Apex has some options to enable features like cross entropy for example. This adds the possibility to enable such variants.
This PR also refactor dependencies and add the latest version of apex.

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-torch-nvidia-apex/package.py [hunk 1]
```diff
@@ -13,44 +13,164 @@ class PyTorchNvidiaApex(PythonPackage, CudaPackage):
...
     def global_options(self, spec, prefix):
         args = []
         if spec.satisfies("^py-torch@1.0:"):
             args.append("--cpp_ext")
-            if "+cuda" in spec:
+            if spec.satisfies("+cuda"):
                 args.append("--cuda_ext")
+
+        if spec.satisfies("+permutation_search_cuda"):
+            args.append("--permutation_search")
```
Comment (msimberg) [inline var/spack/repos/builtin/packages/py-torch-nvidia-apex/package.py:112]: Optional.

Maybe a little helper lambda would be useful here:
```suggestion
        variant_to_arg = lambda v: args.append(f"--{v}") if spec.satisfies(f"+{v}") else None
        variant_to_arg("permutation_search_cuda")
        # etc.
```
Requires that the variant names match the options exactly, unless you want a map between variants and options? Keeping the same names for both might be more straightforward, but you may also have good reasons for adjusting the names, not sure.

Same comment applies to `config_settings`.

Comment (aurianer) [reply var/spack/repos/builtin/packages/py-torch-nvidia-apex/package.py:112]: Yes I thought about that but I don't know why in the README of apex they don't have matching names between name and options. I think i could just decide to use the naming of the setup.py options i.e. the one with `--`. I will update this accordingly

Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/py-torch-nvidia-apex/package.py:112]: As this is an optional comment and there was no other comment, I approve.

**Review by bernhardkaindl — DISMISSED** [review]

**Review by bernhardkaindl — DISMISSED** [review]
@aurianer Approving, and will enable auto-merge with squash.

If you want to cancel auto-merge, close and re-open this PR.

**Review by bernhardkaindl — COMMENTED** [review]
@aurianer Building `torch-nvidia-apex-24.04.01` fails for me on the host that does not have a recent NVidia GPU.
```py
  Warning: Torch did not find available GPUs on this system.
   If your intention is to cross-compile, this is not an error.
  By default, Apex will cross-compile for Pascal (compute capabilities 6.0, 6.1, 6.2),
  Volta (compute capability 7.0), Turing (compute capability 7.5),
  and, if the CUDA version is >= 11.0, Ampere (compute capability 8.0).
  If you wish to cross-compile for a single specific architecture,
  export TORCH_CUDA_ARCH_LIST="compute capability" before running setup.py.



  torch.__version__  = 2.3.1


  Traceback (most recent call last):
    File "/home/t_bernhardk/spack.3/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/py-pip-23.1.2-mykrmrpyjrd7uingef275wiztc3oinm5/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
      main()
    File "/home/t_bernhardk/spack.3/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/py-pip-23.1.2-mykrmrpyjrd7uingef275wiztc3oinm5/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/t_bernhardk/spack.3/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/py-pip-23.1.2-mykrmrpyjrd7uingef275wiztc3oinm5/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 149, in prepare_metadata_for_build_wheel
      return hook(metadata_directory, config_settings)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/t_bernhardk/spack.3/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/py-setuptools-69.2.0-7c5dz5mtkdrd7ggyu6br6vq4adxbfkaf/lib/python3.11/site-packages/setuptools/build_meta.py", line 366, in prepare_metadata_for_build_wheel
      self.run_setup()
    File "/home/t_bernhardk/spack.3/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/py-setuptools-69.2.0-7c5dz5mtkdrd7ggyu6br6vq4adxbfkaf/lib/python3.11/site-packages/setuptools/build_meta.py", line 311, in run_setup
      exec(code, locals())
    File "<string>", line 137, in <module>
    File "<string>", line 24, in get_cuda_bare_metal_version
  TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'
  error: subprocess-exited-with-error
```

@aurianer, can you address this exception at line 24 of get_cuda_bare_metal_version() upstream or add set TORCH_CUDA_ARCH_LIST (e.g. with a variant or so?)

**Review by bernhardkaindl — APPROVED** [review]
@aurianer Due to the merge of #45176, this needed to be rebased on top of the merge of it to develop.

**Comment by aurianer** [comment]
Thank you for rebasing, sorry I was pretty busy these last weeks

**Comment by aurianer** [comment]
@bernhardkaindl I quickly checked the code of the package because I didn't remember what I did and `TORCH_CUDA_ARCH_LIST` is already set through `cuda_arch` did you try installing `py-torch-nvidia-apex+cuda cuda_arch=50` or whichever nvidia GPU you are testing this with? I unfortunately don't have any nvidia GPUs older than Pascal to test that on but setting arch 50 (which is not part of the default architectures according to the error message you posted) works for me :/

**Review by bernhardkaindl — APPROVED** [review]

**Pull request merged**